File: timesheet-example/contracts-java/src/test/java/com/example/contract/InvoiceContractTests.java
package com.example.contract;

import com.google.common.collect.ImmutableList;
import com.example.state.InvoiceState;
import net.corda.core.identity.CordaX500Name;
import net.corda.testing.core.TestIdentity;
import net.corda.testing.node.MockServices;

import java.security.PublicKey;
import java.time.LocalDate;
import java.util.List;
import org.junit.Test;

import static net.corda.testing.node.NodeTestUtils.ledger;


public class InvoiceContractTests {
    private MockServices ledgerServices = new MockServices(ImmutableList.of("com.example.contract"));
    private TestIdentity megaCorp = new TestIdentity(new CordaX500Name("MegaCorp", "London", "GB"));
    private TestIdentity contractor  =new TestIdentity(new CordaX500Name("Contractor", "New York", "US"));
    private TestIdentity oracle = new TestIdentity(new CordaX500Name("Oracle", "New York", "US"));
    private List<PublicKey> signers = ImmutableList.of(megaCorp.getPublicKey(), contractor.getPublicKey(), oracle.getPublicKey());
    private LocalDate date = LocalDate.now();
    private Double rate = 10.0;
    private int invoiceValue = 1;

    @Test
    public void transactionMustIncludeCreateCommand(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                //tx.fails();
                tx.command(signers, new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.verifies();
                return null;
            });
            return null;
        }));
    }

    @Test
    public void transactionMustHaveNoInputs(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.input(InvoiceContract.ID, new InvoiceState(date, invoiceValue,rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.command(signers, new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.failsWith("No inputs should be consumed when issuing an invoice.");
                return null;
            });
            return null;
        }));
    }


    @Test
    public void transactionMustHaveOneOutput(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.command(signers, new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.failsWith("Only one output state should be created.");
                return null;
            });
            return null;
        }));
    }

    @Test
    public void contractorMustSignTransaction(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.command(ImmutableList.of(megaCorp.getPublicKey(),oracle.getPublicKey()), new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.failsWith("All of the participants must be signers.");
                return null;
            });
            return null;
        }));
    }

    @Test
    public void companyMustSignTransaction(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.command(ImmutableList.of(contractor.getPublicKey(),oracle.getPublicKey()), new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.failsWith("All of the participants must be signers.");
                return null;
            });
            return null;
        }));
    }

    @Test
    public void contractorIsNotCompany(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(InvoiceContract.ID, new InvoiceState(date, invoiceValue, rate, megaCorp.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.command(signers, new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.failsWith("The lender and the borrower cannot be the same entity.");
                return null;
            });
            return null;
        }));
    }

    @Test
    public void cannotCreateNegativeValueInvoices(){
        ledger(ledgerServices, (ledger -> {
            ledger.transaction(tx -> {
                tx.output(InvoiceContract.ID, new InvoiceState(date, -1, rate, contractor.getParty(), megaCorp.getParty(), oracle.getParty(), false));
                tx.command(signers, new InvoiceContract.Commands.Create(contractor.getParty(), megaCorp.getParty(), rate));
                tx.failsWith("The Invoice's value must be non-negative.");
                return null;
            });
            return null;
        }));
    }
}

File: timesheet-example/workflows-kotlin/src/test/kotlin/com/example/test/contract/InvoiceContractTests.kt
package com.example.test.contract

import com.example.contract.InvoiceContract
import com.example.state.InvoiceState
import net.corda.core.identity.CordaX500Name
import net.corda.testing.core.TestIdentity
import net.corda.testing.node.MockServices
import net.corda.testing.node.ledger
import org.junit.Test
import java.time.LocalDate

class InvoiceContractTests {
    private val ledgerServices = MockServices(listOf("com.example.contract", "com.example.flow"))
    private val megaCorp = TestIdentity(CordaX500Name("MegaCorp", "London", "GB"))
    private val contractor = TestIdentity(CordaX500Name("Contractor", "New York", "US"))
    private val oracle = TestIdentity(CordaX500Name("Oracle", "New York", "US"))
    private val signers = listOf(megaCorp.publicKey, contractor.publicKey, oracle.publicKey)
    private val date = LocalDate.now()
    private val rate = 10.0
    private val invoiceValue = 1

    @Test
    fun `transaction must include Create command`() {
        ledgerServices.ledger {
            transaction {
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                fails()
                command(signers, InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                verifies()
            }
        }
    }

    @Test
    fun `transaction must have no inputs`() {
        ledgerServices.ledger {
            transaction {
                input(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                command(signers, InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                `fails with`("No inputs should be consumed when issuing an invoice.")
            }
        }
    }

    @Test
    fun `transaction must have one output`() {
        ledgerServices.ledger {
            transaction {
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                command(signers, InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                `fails with`("Only one output state should be created.")
            }
        }
    }

    @Test
    fun `contractor must sign transaction`() {
        ledgerServices.ledger {
            transaction {
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                command(listOf(megaCorp.publicKey, oracle.publicKey), InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                `fails with`("All of the participants must be signers.")
            }
        }
    }

    @Test
    fun `company must sign transaction`() {
        ledgerServices.ledger {
            transaction {
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, contractor.party, megaCorp.party, oracle.party))
                command(listOf(contractor.publicKey, oracle.publicKey), InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                `fails with`("All of the participants must be signers.")
            }
        }
    }

    @Test
    fun `contractor is not company`() {
        ledgerServices.ledger {
            transaction {
                output(InvoiceContract.ID, InvoiceState(date, invoiceValue, rate, megaCorp.party, megaCorp.party, oracle.party))
                command(signers, InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                `fails with`("The lender and the borrower cannot be the same entity.")
            }
        }
    }

    @Test
    fun `cannot create negative-value Invoices`() {
        ledgerServices.ledger {
            transaction {
                output(InvoiceContract.ID, InvoiceState(date, -1, rate, contractor.party, megaCorp.party, oracle.party))
                command(signers, InvoiceContract.Commands.Create(contractor.party, megaCorp.party, rate))
                `fails with`("The Invoice's value must be non-negative.")
            }
        }
    }
}